Cleanup of Firewall Rules

M4t7e opened the pull request:
As discussed in https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/discussions/872 here is a PR to cleanup the firewall rules. All incoming rules except ICMP and SSH are removed.

The SSH rule now has an automatic detection of the current public IPv4 address. To update the firewall with a new IP, the normal `terraform apply` will not work if the IP is not allowed by the firewall anymore, as it does some checks beforehand that need to connect to the nodes. As alternative the following command can be used: `terraform apply -target=module.kube-hetzner.hcloud_firewall.k3s`. Maybe there are even better ways, but this is the best I could find for now.

Also, a new variable called `firewall_ssh_source` has been added to easily configure the firewall rule with static IPs or even disable it completely (`firewall_ssh_source = null`). A static configuration will always overrule the dynamic IP detection.

M4t7e commented:
Had a small fight with `terraform fmt` ^^ Now it should be ready.

mysticaltech commented on data.tf:
```diff
@@ -39,3 +39,14 @@ data "hcloud_ssh_keys" "keys_by_selector" {
 }
+
+data "http" "client_public_ipv4" {
+  url = "https://ipv4.icanhazip.com"
```
@M4t7e  What do you think of @maggie44's comment, she has a point huh? We need to make it compatible with both IPv4 and IPv6, thankfully it seems https://seeip.org/ support both.

mysticaltech replied:
Wow! Super interesting, I believe you. 

@M4t7e Just invited you to join our team, basically you get commit access, you'll be able to create branches, and get to participate more actively in the project when you feel like it. We also have a small discord (not always active, but there when needed). Not all members are always active, so there's no commitment needed. It would be an honor if you accepted to join us 🙏

M4t7e replied:
@mysticaltech Thank you very much, I appreciate that!
I just accepted you invitation :slightly_smiling_face:

mysticaltech replied:
@M4t7e Very good to hear that, welcome to the team!! Here the discord invite https://discord.gg/e56qucwu (will delete tomorrow).

mysticaltech reviewed (commented):
@M4t7e Looking good, just small tweaks and we should be good to go! 🙏🚀

mysticaltech commented on README.md:
````diff
@@ -143,6 +143,32 @@ _Once you start with Terraform, it's best not to change the state of the project
+### Connect via Kube API
+
+Add an additional firewall rule to your kube.tf to access the Kube API. Be careful when exposing the Kube API. It is recommended not to expose it to the public world. If possible, only allow connections from trusted source IPs. Example configuration:
+```hcl
+extra_firewall_rules = [
+  {
+    description = "Allow Incoming Requests to Kube API Server"
+    direction   = "in"
+    protocol    = "tcp"
+    port        = "6443"
+    source_ips  = ["1.2.3.4/32"]
+  }
+]
````
We can uncomment the rule in kube.tf.example, so that it sets it closes it for new users automatically, while old users world same way as before with the default. I think that's the best way to get the best of both worlds.

M4t7e replied:
I could also make it a "managed firewall rule" comparable to the ssh rule in this PR. Then we can set the defaults in a way that users with an existing kube.tf file will see no difference. Something like `firewall_kube_api_source`. What do you think?

mysticaltech replied:
@M4t7e I think that's perfect with a default of  `firewall_kube_api_source = ["0.0.0.0/0", "::/0"]` 🚀

Sneer replied:
@M4t7e I love what you did here with this cleanup. I can only add that `firewall_ssh_source` is great for me (I am using Wireguard on one VPS and restrict all important traffic to only this IP). So this option was a missing thing for me. Also `firewall_kube_api_source ` looks great.

I have been thinking about how to solve the problem of secure connection to the cluster. These four solutions came to my mind.
1) Filtering by IP, the end user has to make sure they have a fixed IP themselves and add IP to `firewall_kube_api_source`
2) Using Netmaker with the Egress Gateway function with a VPS with a fixed IP
3) Using Netmaker together with rootles pods:
a) onetun
b) wireproxy
And here, we do not have any ports open at all, because with the right configuration all we need is ingress in Netmaker and 3 different ports as egress to the kube api.
4) Using Netmaker at the VPS control-plane level: https://docs.netmaker.org/netclient.html#installation and https://itnext.io/multi-cluster-kubernetes-networking-with-netmaker-bfa4e22eb2fb



When I was describing the above solutions I remembered another possibility:
5) Cloudflare tunnel: https://itnext.io/multi-cluster-kubernetes-networking-with-netmaker-bfa4e22eb2fb

Solution 1 is the simplest and therefore probably the most suitable.

M4t7e replied:
Thanks for your input @Sneer! :slightly_smiling_face:

I will finish this PR with configurable `firewall_ssh_source` and `firewall_kube_api_source` for now, but let's think about the more advanced solutions. I would suggest starting a new discussion thread about it and see if we find a good way.

M4t7e commented on README.md:
```diff
@@ -143,6 +143,32 @@ _Once you start with Terraform, it's best not to change the state of the project
+
+### Connect via SSH
+
+Connect to one of the control plane nodes via SSH with `ssh root@<cp-ipv4-address>`. Now you can use kubectl to manage your workloads right away. By default, the firewall only allows SSH connections from your public IPv4 address unless `firewall_ssh_source` is set to a different value in your kube.tf.
```
I can add more context to the Usage chapter about both options.

M4t7e commented on README.md:
```diff
@@ -143,6 +143,32 @@ _Once you start with Terraform, it's best not to change the state of the project
+
+Connect to one of the control plane nodes via SSH with `ssh root@<cp-ipv4-address>`. Now you can use kubectl to manage your workloads right away. By default, the firewall only allows SSH connections from your public IPv4 address unless `firewall_ssh_source` is set to a different value in your kube.tf.
+
+To update your current public IPv4 address to the firewall, you can run:
```
I think here we have a clash between declarative (terraform) and imperative (user wants to change something dynamically) config changes. If we want to support that, I would tend to outsource that functionality from the main terraform based deployment.

Maybe something like that?
- `terraform` apply -> make sure it can connect (open SSH for 0.0.0.0/0 or current public IP) and close it afterwards
- `open-firewall` (terraform, bash, etc.) -> add a new firewall rule (different/dedicated rule from what terraform uses) for 0.0.0.0/0 or current public IP
- `close-firewall` (terraform, bash, etc.) -> delete the firewall rule created by open-firewall

M4t7e replied:
Okay, I see. I think in general that could work with the separated approach. In your CI pipeline you could execute `open-firewall` first (your public IP is whitelisted), apply your Kubernetes changes and execute `close-firewall` after that. But also here are some issues, maybe not in your particular use case, but in general. If you connect in parallel from different pipelines or even as user, multiple instances will try to add and delete the same rule. This approach opens pandora's box of race conditions...

mysticaltech replied:
@M4t7e Absolutely, we already started with that route, with createkh & cleanupkh, but what you are suggesting open-firewall (or open-ssh-port) is needed. I can take care of it, replace all of these features, in one small cli, that works the same way createkh works, and always up to date.

mysticaltech commented:
@M4t7e Fully agree with the above, I think we must prioretize ease of use, while giving the possibility, and showing the way, for being max secure. So `firewall_ssh_source` and `firewall_kube_api_source` opened up to the word is great, but now they can be more secure if they wanted to! That's awesome.

M4t7e commented:
I removed the IP detection for `firewall_ssh_source`, added `firewall_kube_api_source` and made some minor adjustments (descriptions, examples, etc.).

mysticaltech commented:
@maggie44 Now it's up to the user. But firewall rules can be changed even after cluster deploy. You just edit the variable and apply again, it will update via terraform the firewall state directly. No need to ever touch the Hetzner UI.

mysticaltech commented:
@M4t7e Released in v2.4.0, awesome improvement to make the project way more secure! 🥳

codeagencybe commented:
I'm no expert at this, but wouldn't it in general be better to block direct ssh access anyway? 
There are other and better ways to give secure and temporary access to nodes if necessary. 
I'm using Teleport myself and this works amazing. They also handle access via passwordless like Yubikey etc...

They also do features like session recording (and playback) so you know which user did what, they have support for session control, so you can hand out temporary access for eg 1 hour or 1 day or 1 week and it will revoke the access automatically etc...
If you need somebody to troubleshoot specific and temporary, this is an amazing feature. 
And it alles comes in their OSS version. 

ssh access: https://goteleport.com/ssh-server-access/
Kubernetes access: https://goteleport.com/kubernetes-access/

Just my 5 cents, but I would keep this kind of ssh access outside and leave it to something dedicated like Teleport.

mysticaltech commented:
Thanks for sharing @codeagencybe. Now you can block it one flag away, and we needed it to be backward compatible for folks upgrading.

mysticaltech commented:
@maggie44 Cloudflare Zero is amazing, however I did not know how to use it with Kube. Thanks for sharing! That seems perfectly legit, I'll probably use it myself.

codeagencybe commented:
@maggie44 

Are you using the free Cloudflare zero plan or the "pay as you go" plan at 7$/user? 

I was lurking at Headscale (open source variant on Tailscale) to implement zero trust vpn, they also have a lot Kubernetes related parts in the open source repo. 
https://headscale.net/

But this is obviously more work than just settling on an "out of the box" product like Cloudflare zero. 
I like the mindset more of "keeping everything under own control". I have had too many bad experiences in the past years where closed source tools suddenly ramp up the pricing like crazy or shutdown randomly.

mysticaltech commented:
Super good to hear @maggie44, please if you are able to contribute an example setup of that in the examples folder, it would be awesome!

@codeagencybe You definitely have a point, I think the same, but the only ones I trust is cloudflare. While the CEO and CTO remain the same, I think we're in good hands. Honestly I haven't see a service so efficient and well priced as them in the cloud ever. And if they were to phase out, they usually inform well in advance. Probably a lot less work than maintaining a custom in-house open-source based solution. 

And if I am not mistaken, cloudflare zero is free for teams up to 50 people!

codeagencybe commented:
@mysticaltech 

Yes you are right also. In the case of Cloudflare I don't have many doubts about sustainability from their end. 
They have very solid products, and many even free (CDN, protection, DNS, R2 buckets, webworkers, edge, ...)
It's even crazy to understand how they can cope up with that completely for free. 

But there are so many options and companies in the market, especially in niche products that I'm always hesitant and 10x checking should I "bet" my money or business model on this or that. 
Because if it goes wrong, I don't want my entire infrastructure and model to fail apart because of 1 "shady" component in the mix. 
Its a different story if you are just hosting yourself for fun inhouse versus hosting clients critical business software. 

In the case of Cloudflare, even the paid plan at 7$ is peanuts. It's a whole different story if the pricing would be somewhere 800$+/month or a high price + a small uncertain company to rely on.
